gradio/utils.py

```diff
@@ -518,6 +518,32 @@ def resolve_singleton(_list: list[Any] | Any) -> Any:
         return _list
 
 
+def get_all_components() -> list[type[Component] | type[BlockContext]]:
+    import gradio as gr
+
+    classes_to_check = (
+        gr.components.Component.__subclasses__()
+        + gr.blocks.BlockContext.__subclasses__()  # type: ignore
+    )
+    subclasses = []
+
+    while classes_to_check:
+        subclass = classes_to_check.pop()
+        children = subclass.__subclasses__()
+        if children:
+            classes_to_check.extend(children)
```

Comment:
nit: slightly more concise (since if the subclass doesn't have any subclasses, children will be an empty list and extending by an empty list leaves `classes_to_check` unchanged) 
```suggestion
        classes_to_check.extend(subclass.__subclasses__())
```